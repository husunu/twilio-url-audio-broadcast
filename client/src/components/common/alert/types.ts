export enum AlertType {
  SUCCESS = 'success',
  INFO = 'info',
  ERROR = 'error',
  WARNING = 'warning',
}

export type AlertDto = {
  message?: string;
  type?: AlertType;
};
