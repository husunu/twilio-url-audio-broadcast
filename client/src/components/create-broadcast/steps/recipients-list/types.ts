import { INVALID_NUMBERS, VALID_NUMBERS } from './constants';

export type ValidatedNumber = {
  line: number;
  number: string;
  issue?: string | null;
};

export type ValidatedNumbers = {
  [INVALID_NUMBERS]: ValidatedNumber[];
  [VALID_NUMBERS]: ValidatedNumber[];
};
