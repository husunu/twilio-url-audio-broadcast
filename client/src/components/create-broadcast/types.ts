import React from 'react';
import { PropTypes } from '@material-ui/core';

import { StepType, ConfigureStep, RecipientsStep, ReviewStep } from '@/types';
import { ValidatedNumber } from '@/components/create-broadcast/steps/recipients-list/types';

export type StepItem = {
  key: StepType;
  label: string;
  icon: React.ReactNode;
  location: string;
  component: React.FunctionComponent<StepProps>;
  next?: string;
  prev?: string;
  nextButtonColor?: PropTypes.Color;
  nextButtonText?: string;
  nextButtonDisclaimer?: string;
};

export type StepProps = {
  data: ConfigureStep | RecipientsStep | ReviewStep;
  step: StepItem;
};

type NumbersTable = {
  headers: string[];
  rows: Array<keyof ValidatedNumber>;
};

export const VALID_NUMBERS_TABLE: NumbersTable = {
  headers: ['Line', 'Recipients Phone Number'],
  rows: ['line', 'number'],
};

export const INVALID_NUMBERS_TABLE: NumbersTable = {
  headers: ['Line', 'Number', 'Issue'],
  rows: ['line', 'number', 'issue'],
};
