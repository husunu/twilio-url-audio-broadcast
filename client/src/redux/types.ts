import {
  COMPLETE_CONFIGURE_STEP,
  COMPLETE_RECIPIENTS_STEP,
  COMPLETE_REVIEW_STEP,
  DISCARD_BROADCAST,
} from './constants';

import { ConfigureStep, RecipientsStep, ReviewStep } from '@/types';

export type CompleteConfigureStep = {
  type: typeof COMPLETE_CONFIGURE_STEP;
  payload: ConfigureStep;
};

export type CompleteRecipientsStep = {
  type: typeof COMPLETE_RECIPIENTS_STEP;
  payload: RecipientsStep;
};

export type CompleteReviewStep = {
  type: typeof COMPLETE_REVIEW_STEP;
  payload: ReviewStep;
};

export type Steps = {
  [key: string]: ConfigureStep | RecipientsStep | ReviewStep;
};

export type CreateBroadcastState = {
  steps: Steps;
};

export type DiscardBroadcast = {
  type: typeof DISCARD_BROADCAST;
  payload?: any;
};

export type CreateBroadcastActionType =
  | CompleteConfigureStep
  | CompleteRecipientsStep
  | CompleteReviewStep
  | DiscardBroadcast;
