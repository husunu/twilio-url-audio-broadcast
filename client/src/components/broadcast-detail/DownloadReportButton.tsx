import React, { useCallback } from 'react';
import { Button, CircularProgress } from '@material-ui/core';

import { Broadcast, Call, Count } from '@/types';
import { useAllRecipients } from './hooks/useAllRecipients';

type DownloadReportButtonProps = {
  broadcast: Broadcast;
  meta?: Count;
  setHasFailed: (error: boolean) => void;
};

export const DownloadReportButton: React.FC<DownloadReportButtonProps> = ({
  broadcast,
  meta,
  setHasFailed,
}) => {
  const {
    get: getAllRecipients,
    loading,
    error,
    response,
  } = useAllRecipients();

  const onClickDownloadReport = useCallback(async () => {
    setHasFailed(false);
    const { recipients } = await getAllRecipients();

    if (response.ok) {
      const data = {
        id: broadcast.broadcastId,
        name: broadcast.friendlyName,
        message: broadcast.message,
        from: broadcast.from,
        canceled: broadcast.canceled,
        dateCreated: broadcast.dateCreated,
        dateEnd: broadcast.endDate,
        totalRecipients: recipients.length,
        summary: {
          callQueued: meta?.queued,
          callRinging: meta?.ringing,
          callInProgress: meta?.canceled,
          callCanceled: meta?.canceled,
          callCompleted: meta?.completed,
          callFailed: meta?.failed,
          callBusy: meta?.busy,
          callNoAnswer: meta?.['no-answer'],
        },
        recipients: recipients.map((recipient: Call) => ({
          callSid: recipient.callSid,
          number: recipient.to,
          status: recipient.status,
        })),
      };

      const fileName = `notification_report_${broadcast.broadcastId}`;
      const json = JSON.stringify(data, null, 2);
      const blob = new Blob([json], { type: 'application/json' });
      const href = URL.createObjectURL(blob);
      const link = document.createElement('a');

      link.href = href;
      link.download = fileName;

      document.body.appendChild(link);
      link.click();
      document.body.removeChild(link);
    } else if (error) {
      setHasFailed(true);
    }
  }, [
    setHasFailed,
    getAllRecipients,
    response.ok,
    error,
    broadcast.broadcastId,
    broadcast.friendlyName,
    broadcast.message,
    broadcast.from,
    broadcast.canceled,
    broadcast.dateCreated,
    broadcast.endDate,
    meta,
  ]);

  return (
    <Button variant="contained" color="primary" onClick={onClickDownloadReport}>
      Download Report
      {loading && (
        <CircularProgress
          color="secondary"
          size={16}
          style={{ marginLeft: '8px' }}
        />
      )}
    </Button>
  );
};
