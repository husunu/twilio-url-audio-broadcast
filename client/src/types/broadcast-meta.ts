import { CallStatus } from './call-status';

export type Count = {
  queued: number;
  initiated: number;
  ringing: number;
  'in-progress': number;
  completed: number;
  busy: number;
  'no-answer': number;
  canceled: number;
  failed: number;
};

export type BroadcastMeta = {
  total?: number;
  completed?: number;
  count?: Count;
  broadcastStatus?: CallStatus;
};
