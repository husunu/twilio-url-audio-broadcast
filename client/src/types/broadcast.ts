import { Channel } from './channel';

export type Broadcast = {
  broadcastId: string;
  friendlyName: string;
  message: string;
  from: string;
  channel: Channel;
  dateCreated: Date;
  canceled: boolean;
  completed: boolean;
  endDate?: Date;
};
