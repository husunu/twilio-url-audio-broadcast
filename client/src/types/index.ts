export * from './call';
export * from './broadcast';
export * from './broadcast-meta';
export * from './call-status';
export * from './channel';
export * from './step-type';
export * from './steps';
export * from './setter';
