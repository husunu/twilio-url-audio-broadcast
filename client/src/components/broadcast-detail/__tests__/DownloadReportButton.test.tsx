import React from 'react';
import { act, cleanup, render } from '@testing-library/react';

import { DownloadReportButton } from '../DownloadReportButton';
import { CallStatus, Channel } from '@/types';
import { FetchMock } from 'jest-fetch-mock/types';

jest.mock('react-router-dom', () => ({
  useParams: () => ({
    broadcastId: '0001',
  }),
}));

let props = {
  broadcast: {
    broadcastId: '0001',
    friendlyName: 'Test name',
    message: 'Test message',
    from: '+1000001',
    channel: Channel.VOICE,
    dateCreated: new Date(),
    canceled: false,
    completed: true,
    endDate: undefined,
  },
  setHasFailed: jest.fn(),
};

const fetch = global.fetch as FetchMock;

describe('DownloadReportButton', () => {
  beforeEach(async () => {
    fetch.mockResponse(
      JSON.stringify({
        recipients: [
          {
            callSid: 'C000001',
            to: '+1000002',
            status: CallStatus.COMPLETED,
          },
          {
            callSid: 'C000002',
            to: '+1000003',
            status: CallStatus.FAILED,
          },
        ],
      })
    );
  });

  afterEach(() => {
    cleanup();
    fetch.resetMocks();
  });

  test('renders correctly', async () => {
    await act(async () => {
      const { asFragment } = await render(<DownloadReportButton {...props} />);
      expect(asFragment()).toMatchSnapshot();
    });
  });
});
