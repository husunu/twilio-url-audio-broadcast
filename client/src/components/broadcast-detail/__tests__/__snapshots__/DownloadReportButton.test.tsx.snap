// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DownloadReportButton renders correctly 1`] = `
<DocumentFragment>
  <button
    class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
    tabindex="0"
    type="button"
  >
    <span
      class="MuiButton-label"
    >
      Download Report
      <div
        class="MuiCircularProgress-root MuiCircularProgress-colorSecondary MuiCircularProgress-indeterminate"
        role="progressbar"
        style="width: 16px; height: 16px; margin-left: 8px;"
      >
        <svg
          class="MuiCircularProgress-svg"
          viewBox="22 22 44 44"
        >
          <circle
            class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate"
            cx="44"
            cy="44"
            fill="none"
            r="20.2"
            stroke-width="3.6"
          />
        </svg>
      </div>
    </span>
  </button>
</DocumentFragment>
`;
